nelox/scanner.py

```diff
@@ -0,0 +1,117 @@
+from typing import List
+
+from .nelox_token import Token
+from .token_type import TokenType
+
+
+def is_alpha_numeric(c):
+    return is_alpha(c) or is_digit(c)
+
+
+def is_alpha(c):
+    return c.isalpha() or c == '_'
+
+
+def is_digit(c):
+    return '0' <= c <= '9'
+
+
+class Scanner:
+
+    def __init__(self, source: str):
+        self.source: str = source
+        self.tokens: List[Token] = []
+        self.start: int = 0
+        self.current: int = 0
+        self.line: int = 1
+
+    def scan_tokens(self) -> List[Token]:
+        while not self.is_at_end():
+            self.start = self.current
+            token = self.next_token()
+            if token:
+                self.tokens.append(token)
+
+        self.tokens.append(Token(TokenType.EOF, "", None, self.line))
+        return self.tokens
+
+    def is_at_end(self) -> bool:
+        return self.current >= len(self.source)
+
+    def advance(self) -> str:
+        c = self.source[self.current]
+        self.current += 1
+        return c
+
+    def add_token(self, type: TokenType, literal=None) -> Token:
+        text = self.source[self.start:self.current]
+        return Token(type, text, literal, self.line)
+
+    def next_token(self) -> Token | None:
+        c = self.advance()
+
+        if c == '(':
+            return self.add_token(TokenType.LEFT_PAREN)
+        elif c == ')':
+            return self.add_token(TokenType.RIGHT_PAREN)
+        elif c.isdigit():
+            return self.number()
+        elif is_alpha(c):
+            return self.identifier()
+        elif c == '"':
+            return self.string()
+        elif c in (' ', '\r', '\t'):
+            return None
+        elif c == '\n':
+            self.line += 1
+            return None
```

Comment:
I don't think this is the behaviour you want - the lexer should only return None if there's no token to be created (and even then, I think you've chosen to return an EOF token, which is also fine).